### PR TITLE
whats popular section

### PR DESCRIPTION
**Pull request by dima-diachkov**
Create "what's Popular" section [JIRA](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=3371&projectKey=KH2WEB&view=detail&selectedIssue=KH2WEB-5)

cards should include (tv only):
1) photo name, date, rating of the movie
_____________________________________________________________________________________
imageto watch https://dima-diachkov.github.io/copy/

### CODE REVIEW

File: src/App.test.tsx [hunk 1]
```diff
@@ -2,8 +2,7 @@ import React from "react";
 import { render, screen } from "@testing-library/react";
 import App from "./App";
 
-test("renders learn react link", () => {
-  render(<App />);
-  const linkElement = screen.getByText(/What to watch/i);
-  expect(linkElement).toBeInTheDocument();
+test("should render App", () => {
+  render(<App/>);
+  screen.debug();
```
Comment (kerrimov) [inline src/App.test.tsx:7]: test if some element is rendered

File: src/components/CardContainer/CardContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    );
+  }
+
+  return (
+    <Box sx={{ mx: "auto", mt: 10, maxWidth: 1300, flexGrow: 1 }}>
+      <Grid justifyContent="space-around" container>
+        {cardTemplates}
+      </Grid>
+    </Box>
+  );
```
Comment (kerrimov) [inline src/components/CardContainer/CardContainer.tsx:23]: create a mocks folder and move an array with data inside it. After use map here

File: src/components/FilmCard/FilmCard.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+          sx={{
+            "&:hover": {
+              color: "primary.main"
+            }
+          }}>
+          <Typography component="h3" sx={{fontSize: "1.1rem", fontWeight: 700}}>
+            {"American siege"}
+          </Typography>
+        </Link>
+      </CardActions>
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/FilmCard/FilmCard.tsx:52]: done

File: src/components/FilmCard/FilmCard.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import * as React from "react";
+import Card from "@mui/material/Card";
+import CardMedia from "@mui/material/CardMedia";
+import { Link, CardActions, Grid, Typography, CardActionArea, Checkbox } from "@mui/material";
+import CircularProgressWithLabel from "../RaitingLabel/RaitingLabel";
+import FavoriteBorder from "@mui/icons-material/FavoriteBorder";
+import Favorite from "@mui/icons-material/Favorite";
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/FilmCard/FilmCard.tsx:7]: done

File: src/components/RaitingLabel/RaitingLabel.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+          display: "flex",
+          alignItems: "center",
+          justifyContent: "center",
+        }}
+      >
+        <Typography
+          variant="caption"
+          component="div"
+          color="white"
+        >{`${Math.round(props.value)}%`}</Typography>
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/RaitingLabel/RaitingLabel.tsx:34]: formatting fixed

File: src/components/RaitingLabel/RaitingLabel.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        sx={{
+          top: 0,
+          left: 0,
+          bottom: 0,
+          right: 0,
+          position: "absolute",
+          display: "flex",
+          alignItems: "center",
+          justifyContent: "center",
+        }}
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/RaitingLabel/RaitingLabel.tsx:28]: done

File: src/App.test.tsx (hunk 1 above)
Comment (kerrimov) [reply src/App.test.tsx:9]: it is a good idea to add a new test, but you need to update this one

File: src/mocks/cardData.tsx [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import FilmCard from "../components/FilmCard/FilmCard";
+
+const cardTemplates: Array<React.ReactElement> = [];
+for (let i = 0; i < 7; i++) {
+    cardTemplates.push(
+        <Grid item xs={6} sm={4} md={3} lg={1.7} key={i}>
+            <FilmCard />
+        </Grid>
+    ); 
+}
```
Comment (kerrimov) [inline src/mocks/cardData.tsx:12]: Mocks are just data in an array or JSON-like format. It is not a component. So we create mocks as we expect data from backend in the future.

File: src/App.test.tsx [hunk 8]
```diff
@@ -1,9 +1,22 @@
...
+test("check label", ()=> {
+  const { getByText } = render(<RaitingLabel value={55}/>);
+  const label = getByText(/55%/i);
+  expect(label).toBeTruthy();
+});
+
+test("check film name", () => {
+  render(<CardInfoSection/>);
+  const filmName = screen.getByText(/American siege/i);
+  expect(filmName).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/App.test.tsx:21]: done

File: src/components/CardContainer/CardContainer.tsx [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import * as React from "react";
+import Box from "@mui/material/Box";
+import Grid from "@mui/material/Grid";
+import cardTemplates from "../../mocks/cardData";
+
+export default function BasicGrid() {
+
+  return (
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardContainer/CardContainer.tsx:8]: done

File: src/components/CardInfoSection/CardInfoSection.tsx [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        <CardActions sx={{ minHeight: 10, pt: 0, pb: 0 }}>
+            <Link
+                href="#"
+                underline="none"
+                color="text.primary"
+                sx={{
+                    "&:hover": {
+                        color: "primary.main"
+                    }
+                }}>
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardInfoSection/CardInfoSection.tsx:15]: done

File: src/components/FilmCard/FilmCard.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            checkedIcon={<Favorite />}
+            sx={{ p: 0.5 }}
+            title="add to favorite"
+          />
+        </Grid>
+        <CardMedia
+          sx={{ borderRadius: 3, mt: -4 }}
+          component="img"
+          height="225"
+          image="https://www.themoviedb.org/t/p/w440_and_h660_face/daeVrgyj0ue8qb3AHyU3UeCwoZz.jpg"
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/FilmCard/FilmCard.tsx:24]: image moved

File: src/components/RaitingLabel/RaitingLabel.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  display: "flex",
+  alignItems: "center",
+  justifyContent: "center",
+};
+
+interface Props {
+  value: number;
+}
+
+export default function CircularProgressWithLabel(props: Props) {
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/RaitingLabel/RaitingLabel.tsx:21]: done

File: src/components/CardInfoSection/CardInfoSection.tsx [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+                <Typography component="h3" sx={{ fontSize: "1.1rem", fontWeight: 700 }}>
+                    {"American siege"}
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardInfoSection/CardInfoSection.tsx:17]: really don't need

File: src/components/CardInfoSection/CardInfoSection.tsx [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import * as React from "react";
+import { CardActions, Link, Typography } from "@mui/material";
+
+const CardInfoSection = () => {
+    return (
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardInfoSection/CardInfoSection.tsx:5]: without return, the app works, but the test needs the component to return the JSX element

image

File: src/components/FilmCard/FilmCard.tsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        <Grid container justifyContent="flex-end" padding={0}>
+          <Checkbox
+            icon={<FavoriteBorder sx={{ color: "primary.main" }} />}
+            checkedIcon={<Favorite />}
+            sx={{ p: 0.5 }}
+            title="add to favorite"
+          />
+        </Grid>
+        <CardMedia
+          sx={{ borderRadius: 3, mt: -4 }}
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/FilmCard/FilmCard.tsx:21]: mt: 0
image
mt: -4
image

File: src/mocks/cardData.tsx (hunk 7 above)
Comment (dima-diachkov) [reply src/mocks/cardData.tsx]: crystal clear. done.

File: src/components/CardInfoSection/CardInfoSection.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (kerrimov) [reply src/components/CardInfoSection/CardInfoSection.tsx]: your syntax should be like `CardInfoSection = () => ()`, so jsx element will be retuned

File: src/mocks/cardData.json [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        "default-film-name": "The Batman",
+        "default-film-date": "01 Mar 2022",
+        "film-rating": 78
+    },
+    {
+        "default-image-for-card": "https://www.themoviedb.org/t/p/w220_and_h330_face/hangTmbxpSV4gpHG7MgSlCWSSFa.jpg",
+        "default-film-name": "The Exorcism of God",
+        "default-film-date": "10 Feb 2022",
+        "film-rating": 68
+    },
```
Comment (kerrimov) [inline src/mocks/cardData.json:25]: please change this film to `The Intouchables`, 
link: https://image.tmdb.org/t/p/w220_and_h330_face/323BP0itpxTsO0skTwdnVmf7YC9.jpg

File: src/components/FilmCard/FilmCard.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (kerrimov) [reply src/components/FilmCard/FilmCard.tsx:27]: something wrong with markup, using minus is not a good idea.

File: src/components/FilmCard/FilmCard.tsx [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+interface Props {
+  image: string;
+  name: string;
+  date: string;
+  rating: number;
+}
+
+export const FilmCard = ({ image, name, date, rating }: Props) => {
+  return (
```
Comment (kerrimov) [inline src/components/FilmCard/FilmCard.tsx:15]:
```suggestion
export const FilmCard = ({ image, name, date, rating }: Props) => (
```

File: src/components/CardContainer/CardContainer.tsx [hunk 18]
```diff
@@ -0,0 +1,23 @@
+import * as React from "react";
+import { Box, Grid } from "@mui/material";
+import cardData from "../../mocks/cardData.json";
+import { FilmCard } from "../FilmCard/FilmCard";
+
+export const CardContainer = () => {
+  return (
```
Comment (kerrimov) [inline src/components/CardContainer/CardContainer.tsx:7]: the same

File: src/App.test.tsx (hunk 1 above)
Comment (dima-diachkov) [reply src/App.test.tsx]: updated

File: src/tests/RatingLabel.test.tsx [hunk 19]
```diff
@@ -0,0 +1,9 @@
+import React from "react";
+import { render } from "@testing-library/react";
+import { CircularProgressWithLabel } from "../components/RaitingLabel/RaitingLabel";
+
+test("check label", ()=> {
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/tests/RatingLabel.test.tsx]: done

File: src/mocks/cardData.json [hunk 20]
```diff
@@ -0,0 +1,44 @@
+[
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/mocks/cardData.json:1]: transfer to cardData.ts

File: src/mocks/cardData.json (hunk 16 above)
Comment (dima-diachkov) [reply src/mocks/cardData.json:25]: scary... done

File: src/components/CardInfoSection/CardInfoSection.tsx [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+        color: "primary.main"
+    }
+};
+
+interface Props {
+    name: string;
+}
+
+export const CardInfoSection = ({ name }: Props) => {
+    return (
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardInfoSection/CardInfoSection.tsx:15]: done

File: src/components/CardInfoSection/CardInfoSection.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardInfoSection/CardInfoSection.tsx]: done

File: src/components/CardContainer/CardContainer.tsx [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    <Box sx={{ mx: "auto", mt: 10, maxWidth: 1300, flexGrow: 1 }}>
+      <Grid justifyContent="space-around" container>
+        {cardData.map((card, index) => {
+          return <Grid item xs={6} sm={4} md={3} lg={1.7} key={index}>
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardContainer/CardContainer.tsx:11]: done

File: src/components/FilmCard/FilmCard.tsx (hunk 17 above)
Comment (dima-diachkov) [reply src/components/FilmCard/FilmCard.tsx:15]: done

File: src/components/CardContainer/CardContainer.tsx (hunk 18 above)
Comment (dima-diachkov) [reply src/components/CardContainer/CardContainer.tsx:7]: done

File: src/components/CardContainer/CardContainer.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/CardContainer/CardContainer.tsx:11]: it's not works here(

File: src/components/RaitingLabel/RaitingLabel.tsx [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+export const CircularProgressWithLabel = ({value}: Props) => {
+  return (
+    <Box sx={{
+      position: "relative",
+      display: "inline-flex",
+      backgroundColor: "black",
+      borderRadius: "50%"
+    }}
+    >
+      <CircularProgress variant="determinate" {...{value}} />
```
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/RaitingLabel/RaitingLabel.tsx]: it's not spread it's rest. there's syntax for this mui component. ok, I replaced to value={value} ...still works

File: src/components/FilmCard/FilmCard.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (dima-diachkov) [reply src/components/FilmCard/FilmCard.tsx]: I read about ways to solve this problem and redid it with `position: "relative" position: "absolute"`. there are no "minuses", but for me, the solution is not something better and understandable. 
there are points of view both for
https://stackoverflow.com/questions/8684917/is-it-bad-practice-to-use-negative-margins-or-padding-in-css
 and against
https://www.quora.com/What-are-your-opinion-on-negative-margin-in-CSS

File: src/mocks/cardData.ts [hunk 24]
```diff
@@ -0,0 +1,44 @@
+export const cardData = [
```
Comment (kerrimov) [inline src/mocks/cardData.ts:1]: add interface for mocks

Comment (dima-diachkov) [reply src/mocks/cardData.ts:1]: done

File: src/components/CardContainer/CardContainer.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (kerrimov) [reply src/components/CardContainer/CardContainer.tsx:11]: try again. discussed on daily

File: src/mocks/cardData.ts [hunk 25]
```diff
@@ -0,0 +1,51 @@
+interface Card {
+    "default-image-for-card": string;
+    "default-film-name": string;
+    "default-film-date": string;
+    "film-rating": number;
```
Comment (kerrimov) [inline src/mocks/cardData.ts:5]: in what format does the backend return the data?

File: src/index.tsx [hunk 26]
```diff
@@ -1,6 +1,6 @@
+import { ThemeProvider } from "@mui/material";
```
Comment (kerrimov) [inline src/index.tsx:1]: move import to line 3

File: src/mocks/cardData.ts [hunk 27]
```diff
@@ -0,0 +1,51 @@
+interface Card {
```
Comment (kerrimov) [inline src/mocks/cardData.ts:1]: create folder types and move interface

File: src/components/RaitingLabel/RaitingLabel.tsx [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+  display: "flex",
+  alignItems: "center",
+  justifyContent: "center",
+};
+
+interface Props {
+  value: number;
+}
+
+export const CircularProgressWithLabel = ({ value }: Props) => (
```
Comment (kerrimov) [inline src/components/RaitingLabel/RaitingLabel.tsx:19]: Your component should be named as the file. Rename value to a more understandable name.

**Review by kerrimov — COMMENTED** [review]
All components should have an index.ts file. Check PR https://github.com/kerrimov/what-to-watch/pull/5.
And move your components to the parents folder if they are nested.

File: src/index.tsx (hunk 26 above)
Comment (dima-diachkov) [reply src/index.tsx:1]: moved

File: src/components/RaitingLabel/RaitingLabel.tsx (hunk 28 above)
Comment (dima-diachkov) [reply src/components/RaitingLabel/RaitingLabel.tsx:19]: done

File: src/components/CardContainer/CardContainer.tsx [hunk 29]
```diff
@@ -0,0 +1,23 @@
+import * as React from "react";
+import { Box, Grid } from "@mui/material";
+import { FilmCard } from "./FilmCard/FilmCard";
```
Comment (kerrimov) [inline src/components/CardContainer/CardContainer.tsx:3]: import from folder root as you have added index.ts

File: src/components/CardContainer/FilmCard/FilmCard.tsx [hunk 30]
```diff
@@ -0,0 +1,43 @@
+import * as React from "react";
+import { Card, Grid, CardMedia, CardActionArea, Checkbox } from "@mui/material";
+import { FavoriteBorder, Favorite } from "@mui/icons-material";
+import { CardInfoSection } from "./CardInfoSection/CardInfoSection";
+import { RatingLabel } from "./RaitingLabel/RaitingLabel";
```
Comment (kerrimov) [inline src/components/CardContainer/FilmCard/FilmCard.tsx:5]: import from folder root as you have added index.ts

File: src/components/CardContainer/types/Card.ts [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+// interface CardData extends Array<Card> {
+//     [index: number]: Card;
+// }
+
+export type Card = {
+    "default-image-for-card": string;
+    "default-film-name": string;
+    "default-film-date": string;
+    "film-rating": number;
+};
```
Comment (kerrimov) [inline src/components/CardContainer/types/Card.ts:28]: Should be interface with the fields that UI expects from the backend

File: src/mocks/cardData.ts (hunk 25 above)
Comment (kerrimov) [reply src/mocks/cardData.ts]: Should be interface with the fields that UI expects from the backend

File: src/components/CardContainer/FilmCard/FilmCard.tsx (hunk 30 above)
Comment (dima-diachkov) [reply src/components/CardContainer/FilmCard/FilmCard.tsx]: done

File: src/components/CardContainer/CardContainer.tsx (hunk 29 above)
Comment (dima-diachkov) [reply src/components/CardContainer/CardContainer.tsx]: done

File: src/mocks/cardData.ts (hunk 25 above)
Comment (dima-diachkov) [reply src/mocks/cardData.ts]: used field names like in JSON which returns backend

File: src/mocks/cardData.ts [hunk 32]
```diff
@@ -0,0 +1,53 @@
+import { Card } from "../components/CardContainer/types/Card";
+
+export const cardData: Array<Card> /*CardData*/ = [
```
Comment (kerrimov) [inline src/mocks/cardData.ts:3]: remove this comment

Comment (dima-diachkov) [reply src/mocks/cardData.ts:3]: removed

**Review by kerrimov — DISMISSED** [review]
please squash your commits

**Review by kerrimov — APPROVED** [review]

**Review by kerrimov — APPROVED** [review]

**Pull request merged**